### PR TITLE
Changes in parse.py

### PR DESCRIPTION
**Pull request by diivm**
# Objective 1: Change identifier type of the AST.
### Background
For a tree node, we were using `clang_bind.parse.Node` (data holder class) objects as identifiers (`treelib.Tree.identifier`). 

This created a problem while attempting to create subtrees from the tree:
- Subtree creation (and other functions) uses copy functions (from `copy` module) internally, which lead to:
  `ValueError: ctypes objects containing pointers cannot be pickled`
- This is caused because the data holder class had clang pointers (cursors, `clang.cindex.Cursor`) which uses `ctypes` internally. Thus, their copies couldn't be created and subsequently, any tree operations couldn't be performed.

This can be solved by a simple method: instead of using the data holder objects (and thus `ctypes`) as identifiers, we use UUIDs as identifiers and map them to the data holder objects. 
- This won't create any difference in the tree representation as the tree printing operations rely on the `tag`, not the identifier and we are keeping the tag same (`repr` of the data holder).
- The UUID creation process is handled internally by `treelib`, so we just let it create UUIDs and we map them to the data holder in our `clang_bind.parse.Parse` class.

> We also refactor `clang_bind.parse.Node` (data holder) class to `clang_bind.parse.ParsedInfo` as it created confusion while using `treelib.Node` objects in our code

### Corresponding changes in `parse.py` (2b6d961):
- New tree uses uuids (internal to `treelib`) as node identifiers instead of the Node (now ParsedInfo) objects.
- refactor `Node` class to `ParsedInfo`
- Introduced `_parsed_info_map`
- Add some new functions

### Corresponding changes in `test_parse.py` (1a5d5bb):
- New tree has UUIDs as identifiers (internal), as opposed to the `Node` (now `ParsedInfo`) objects

### Other changes in `parse.py` (2b6d961):
- Improved documentation
- Correct private methods' name

---

# Objective 2: Improve structure of tests in `test_parse.py`

### Corresponding changes in `test_parse.py` (47997ef):
- use a class
- remove the hard-coded filename
- remove the dependency of each function on `tmp_path`
- use `tempfile` module instead; file writing method changes
- update all functions to accommodate the new needs

### CODE REVIEW

File: clang_bind/parse.py [hunk 1]
```diff
@@ -38,40 +51,99 @@ def __init__(self, file, compiler_arguments):
...
-    def _construct_tree(self, node):
+        :param child_cursor: The child cursor to check, an object of :class:`clang.cindex.Cursor`
+        :type child_cursor: class:`clang.cindex.Cursor`
+        :return: `True` if child cursor in file, else `False`
+        :rtype: bool
         """
-        Recursively generates tree by traversing the AST of the node.
+        return self.is_cursor_in_file(child_cursor, self.filename)
+
+    def __construct_tree(self, node):
```
Comment (kunaltyagi) [inline clang_bind/parse.py]: Why __construct_tree instead of _construct_tree?

File: clang_bind/parse.py [hunk 2]
```diff
@@ -38,40 +51,99 @@ def __init__(self, file, compiler_arguments):
...
+        , by getting the values from `_parsed_info_map`
+
+        :param node_ids: Node identifiers of :class:`treelib.Node`
+        :type node_ids: list
+        :return: Parsed infos of :class:`clang_bind.parse.ParsedInfo`
+        :rtype: list
+        """
+        return list(map(lambda node_id: self._parsed_info_map.get(node_id), node_ids))
+
+    def get_node_ids_from_nodes(self, nodes):
```
Comment (kunaltyagi) [inline clang_bind/parse.py]: Static or free function?

File: tests/test_parse.py [hunk 3]
```diff
@@ -1,403 +1,404 @@
+import tempfile
```
Comment (kunaltyagi) [inline tests/test_parse.py:1]: Love it

File: tests/test_parse.py [hunk 4]
```diff
@@ -1,403 +1,404 @@
...
-def test_anonymous_decls(tmp_path):
-    file_contents = """
-    union {
-        struct {
-            enum {};
+class TestParse:
+    parser, tree, tree_paths = None, None, None
+    file = tempfile.NamedTemporaryFile(suffix=".cpp")
+
+    def __parse(self, file_contents):
```
Comment (kunaltyagi) [inline tests/test_parse.py]: You seem to love using double underscore. I'd like to know the benefits wrt single underscore, so I don't repeatedly call you out on something that  shouldn't be called out

File: tests/test_parse.py [hunk 5]
```diff
@@ -1,403 +1,404 @@
...
+        function_decl = self.tree_paths[0][1]
+        assert function_decl.cursor.kind == clang.CursorKind.FUNCTION_DECL
+        assert function_decl.cursor.spelling == "aFunction"
+        assert function_decl.cursor.result_type.kind == clang.TypeKind.INT
+
+        function_decl_children = self.parser.get_parsed_infos_from_node_ids(
+            self.parser.get_node_ids_from_nodes(
+                self.parser.get_children_nodes_from_parent_parsed_info(function_decl)
+            )
+        )
```
Comment (kunaltyagi) [inline tests/test_parse.py:135]: This looks gnarly. Helper function to wrap the 3 calls into a single call?

**Review by kunaltyagi — COMMENTED** [review]
I like the new tests.

File: clang_bind/parse.py (hunk 1 above)
Comment (diivm) [reply clang_bind/parse.py]: I thought double underscore was the standard. Changed.

File: tests/test_parse.py (hunk 4 above)
Comment (diivm) [reply tests/test_parse.py]: I thought double underscore was the standard. Changed.

File: tests/test_parse.py (hunk 5 above)
Comment (diivm) [reply tests/test_parse.py:135]: Done.

**Pull request merged**